File: drawing-logic/arrows.py
class ArrowMarker:
    def __init__(self, start_x, start_y, end_x, end_y):
        self.start_x = start_x
        self.start_y = start_y
        self.end_x = end_x
        self.end_y = end_y
        self.adjustment_points = self.calculate_adjustment_points()

    def calculate_adjustment_points(self):
        # Calculate adjustment points (e.g., at the start and end of the arrow)
        return {
            "start": (self.start_x, self.start_y),
            "end": (self.end_x, self.end_y),
        }

    def update_coordinates(self, point, new_x, new_y):
        if point == "start":
            self.start_x = new_x
            self.start_y = new_y
        elif point == "end":
            self.end_x = new_x
            self.end_y = new_y
        self.adjustment_points = self.calculate_adjustment_points()

    def draw(self):
        # Basic arrow marker drawing logic (can be expanded)
        print(f"Drawing Arrow Marker from ({self.start_x}, {self.start_y}) to ({self.end_x}, {self.end_y})")

    def to_dict(self):
        return {
            "type": "arrow_marker",
            "start_x": self.start_x,
            "start_y": self.start_y,
            "end_x": self.end_x,
            "end_y": self.end_y,
            "adjustment_points": self.adjustment_points
        }

class Arrow:
    def __init__(self, start_x, start_y, end_x, end_y):
        self.start_x = start_x
        self.start_y = start_y
        self.end_x = end_x
        self.end_y = end_y
        self.adjustment_points = self.calculate_adjustment_points()

    def calculate_adjustment_points(self):
        # Calculate adjustment points (e.g., at the start and end of the arrow)
        return {
            "start": (self.start_x, self.start_y),
            "end": (self.end_x, self.end_y),
        }

    def update_coordinates(self, point, new_x, new_y):
        if point == "start":
            self.start_x = new_x
            self.start_y = new_y
        elif point == "end":
            self.end_x = new_x
            self.end_y = new_y
        self.adjustment_points = self.calculate_adjustment_points()

    def draw(self):
        # Basic arrow drawing logic (can be expanded)
        print(f"Drawing Arrow from ({self.start_x}, {self.start_y}) to ({self.end_x}, {self.end_y})")

    def to_dict(self):
        return {
            "type": "arrow",
            "start_x": self.start_x,
            "start_y": self.start_y,
            "end_x": self.end_x,
            "end_y": self.end_y,
            "adjustment_points": self.adjustment_points
        }

class ArrowMarkUp:
    def __init__(self, start_x, start_y, end_x, end_y):
        self.start_x = start_x
        self.start_y = start_y
        self.end_x = end_x
        self.end_y = end_y
        self.adjustment_points = self.calculate_adjustment_points()

    def calculate_adjustment_points(self):
        # Calculate adjustment points (e.g., at the start and end of the arrow)
        return {
            "start": (self.start_x, self.start_y),
            "end": (self.end_x, self.end_y),
        }

    def update_coordinates(self, point, new_x, new_y):
        if point == "start":
            self.start_x = new_x
            self.start_y = new_y
        elif point == "end":
            self.end_x = new_x
            self.end_y = new_y
        self.adjustment_points = self.calculate_adjustment_points()

    def draw(self):
        # Basic arrow mark up drawing logic (can be expanded)
        print(f"Drawing Arrow Mark Up from ({self.start_x}, {self.start_y}) to ({self.end_x}, {self.end_y})")

    def to_dict(self):
        return {
            "type": "arrow_mark_up",
            "start_x": self.start_x,
            "start_y": self.start_y,
            "end_x": self.end_x,
            "end_y": self.end_y,
            "adjustment_points": self.adjustment_points
        }

class ArrowMarkDown:
    def __init__(self, start_x, start_y, end_x, end_y):
        self.start_x = start_x
        self.start_y = start_y
        self.end_x = end_x
        self.end_y = end_y
        self.adjustment_points = self.calculate_adjustment_points()

    def calculate_adjustment_points(self):
        # Calculate adjustment points (e.g., at the start and end of the arrow)
        return {
            "start": (self.start_x, self.start_y),
            "end": (self.end_x, self.end_y),
        }

    def update_coordinates(self, point, new_x, new_y):
        if point == "start":
            self.start_x = new_x
            self.start_y = new_y
        elif point == "end":
            self.end_x = new_x
            self.end_y = new_y
        self.adjustment_points = self.calculate_adjustment_points()

    def draw(self):
        # Basic arrow mark down drawing logic (can be expanded)
        print(f"Drawing Arrow Mark Down from ({self.start_x}, {self.start_y}) to ({self.end_x}, {self.end_y})")

    def to_dict(self):
        return {
            "type": "arrow_mark_down",
            "start_x": self.start_x,
            "start_y": self.start_y,
            "end_x": self.end_x,
            "end_y": self.end_y,
            "adjustment_points": self.adjustment_points
        }

File: drawing-java/arrows.js
class ArrowMarker {
    constructor({ startX, startY, endX, endY }) {
        this.startX = startX;
        this.startY = startY;
        this.endX = endX;
        this.endY = endY;
        this.adjustmentPoints = this.calculateAdjustmentPoints();
    }

    calculateAdjustmentPoints() {
        return {
            start: { x: this.startX, y: this.startY },
            end: { x: this.endX, y: this.endY },
        };
    }

    updateCoordinates(point, newX, newY) {
        if (point === 'start') {
            this.startX = newX;
            this.startY = newY;
        } else if (point === 'end') {
            this.endX = newX;
            this.endY = newY;
        }
        this.adjustmentPoints = this.calculateAdjustmentPoints();
    }

    draw(context) {
        context.beginPath();
        context.moveTo(this.startX, this.startY);
        context.lineTo(this.endX, this.endY);
        context.stroke();
    }

    toObject() {
        return {
            type: 'arrow_marker',
            startX: this.startX,
            startY: this.startY,
            endX: this.endX,
            endY: this.endY,
            adjustmentPoints: this.adjustmentPoints
        };
    }
}

class Arrow extends ArrowMarker {
    draw(context) {
        const headLength = 10;
        const angle = Math.atan2(this.endY - this.startY, this.endX - this.startX);
        
        context.beginPath();
        context.moveTo(this.startX, this.startY);
        context.lineTo(this.endX, this.endY);
        
        context.lineTo(
            this.endX - headLength * Math.cos(angle - Math.PI / 6),
            this.endY - headLength * Math.sin(angle - Math.PI / 6)
        );
        context.moveTo(this.endX, this.endY);
        context.lineTo(
            this.endX - headLength * Math.cos(angle + Math.PI / 6),
            this.endY - headLength * Math.sin(angle + Math.PI / 6)
        );
        
        context.stroke();
    }
}

class ArrowMarkUp extends ArrowMarker {
    draw(context) {
        const arrowSize = 8;
        
        context.beginPath();
        context.moveTo(this.startX, this.startY);
        context.lineTo(this.endX, this.endY);
        context.lineTo(this.endX - arrowSize, this.endY + arrowSize);
        context.moveTo(this.endX, this.endY);
        context.lineTo(this.endX + arrowSize, this.endY + arrowSize);
        context.stroke();
    }
}

class ArrowMarkDown extends ArrowMarker {
    draw(context) {
        const arrowSize = 8;
        
        context.beginPath();
        context.moveTo(this.startX, this.startY);
        context.lineTo(this.endX, this.endY);
        context.lineTo(this.endX - arrowSize, this.endY - arrowSize);
        context.moveTo(this.endX, this.endY);
        context.lineTo(this.endX + arrowSize, this.endY - arrowSize);
        context.stroke();
    }
}

export { ArrowMarker, Arrow, ArrowMarkUp, ArrowMarkDown };
